src/librustc/middle/region.rs

```diff
@@ -735,6 +751,8 @@ fn resolve_stmt<'a, 'tcx>(visitor: &mut RegionResolutionVisitor<'a, 'tcx>, stmt:
 fn resolve_expr<'a, 'tcx>(visitor: &mut RegionResolutionVisitor<'a, 'tcx>, expr: &'tcx hir::Expr) {
     debug!("resolve_expr(expr.id={:?})", expr.id);
 
+    visitor.expr_count += 1;
```

Comment:
I find this mechanism somewhat hacky (traversing the CFG in preorder to gather this ordering). It would be nice to at least justify why it works:

 - execution can't go backwards in node postorder, except through a backedge (aka loop)
 - if node A is preorder-before node B, but node B is postorder-before node A, then A is a parent of B. This comes from the definition of a DFS.

Therefore, assuming that expr A is preorder-after a yield, we want to show that the results of expr A can't be used "over" the yield. There are 2 cases:
1. if expr A is postorder-after the yield, then it can only reach the yield across a loop it is in. But a loop is a terminating scope and drops all temporaries created within it.
2. if the yield is postorder-after expr A, then the yield is a parent of expr A. If we make yields a terminating scope (see the borrow-in-yield-expr example), then the terminating scope would destroy the temporary created by expr A.

Reply:
I'm still a bit uncomfortable with the hackiness here.

Reply:
Oops. Assignments are evaluated RTL, so you can do
```Rust
    *{ yield; x } = &y;
```

And `&y` would be stored over the yield.

Maybe use the old CFG and do dataflow over it? That would be ugly, but work.

Reply:
Visitor goes over subexpressions in RTL, so this actually works

Now there's this (archaic in incorrect but correct-in-spirit) comment:
https://github.com/rust-lang/rust/blob/d93036a043dbd19f394ad320313bdbe81644f121/src/librustc/hir/intravisit.rs#L29-L34